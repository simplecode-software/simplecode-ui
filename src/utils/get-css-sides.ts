import createCachedSelector from 're-reselect';

import { TStringList } from '../types';
import { getCssArgs } from './get-css-args';


const getCssSidesN = createCachedSelector<
    string,
    TStringList,
    string,
    TStringList,
    React.CSSProperties | null
>(
    (name) => name,
    (_name, splitted) => splitted,
    (name, splitted) => {
        if (!splitted.length) {
            return null;
        }
        
        const mapped = splitted.map((v) => ((Number.isNaN(+v) ? v : +v) ?? 0));
        
        if (mapped.length === 1) {
            return {
                [`${name}Top`]: mapped[0] ?? 0,
                [`${name}Right`]: mapped[0] ?? 0,
                [`${name}Bottom`]: mapped[0] ?? 0,
                [`${name}Left`]: mapped[0] ?? 0,
            };
        }
        if (mapped.length === 2) {
            return {
                [`${name}Top`]: mapped[0] ?? 0,
                [`${name}Right`]: mapped[1] ?? 0,
                [`${name}Bottom`]: mapped[0] ?? 0,
                [`${name}Left`]: mapped[1] ?? 0,
            };
        }
        if (mapped.length === 3) {
            return {
                [`${name}Top`]: mapped[0] ?? 0,
                [`${name}Right`]: mapped[1] ?? 0,
                [`${name}Bottom`]: mapped[2] ?? 0,
                [`${name}Left`]: mapped[1] ?? 0,
            };
        }

        return {
            [`${name}Top`]: mapped[0] ?? 0,
            [`${name}Right`]: mapped[1] ?? 0,
            [`${name}Bottom`]: mapped[2] ?? 0,
            [`${name}Left`]: mapped[3] ?? 0,
        };
    },
)((name, splitted) => `getCssSides/getCssSides-${splitted.length}-${name}-${splitted.join(',')}`);


export const getCssSides = createCachedSelector<
    string,
    string | number | null,
    string,
    string | number | null,
    React.CSSProperties | null
>(
    (name) => name,
    (_name, value) => value,
    (name, value) => {
        if (!name) {
            return null;
        }
        if (typeof value === 'number') {
            return { [name]: value };
        }
        if (!value) {
            return null;
        }
        
        const splitted: TStringList | null = getCssArgs(value);
        
        if (!splitted) {
            return null;
        }
        
        return getCssSidesN(name, splitted);
    },
)((name, value) => `getCssSides-${name}-${value}`);
