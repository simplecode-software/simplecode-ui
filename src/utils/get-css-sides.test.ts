import { getCssSides } from './get-css-sides';


describe('Get Css Sides', () => {
    it('is a function', () => {
        expect(typeof getCssSides).toBe('function');
    });
    
    it('computes single side', () => {
        expect(getCssSides('a', '10')).toMatchSnapshot();
    });
    
    it('computes double sides', () => {
        expect(getCssSides('b', '10 20')).toMatchSnapshot();
    });
    
    it('computes tripple sides', () => {
        expect(getCssSides('c', '10 20 30')).toMatchSnapshot();
    });
    
    it('computes full sides', () => {
        expect(getCssSides('d', '10 20 30 40')).toMatchSnapshot();
    });
    
    it('trims extra sides', () => {
        expect(getCssSides('e', '10 20 30 40 50')).toMatchSnapshot();
    });
    
    it('computes number padding', () => {
        expect(getCssSides('f', 40)).toMatchSnapshot();
    });
    
    it('returns null if no padding', () => {
        expect(getCssSides('g', null)).toMatchSnapshot();
    });
    
    it('returns null if empty string', () => {
        expect(getCssSides('h', '')).toMatchSnapshot();
    });
    
    it('returns null if the name is empty', () => {
        expect(getCssSides('', 1)).toMatchSnapshot();
    });
});
