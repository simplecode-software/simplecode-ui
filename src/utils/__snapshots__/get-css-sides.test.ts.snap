// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get Css Sides computes double sides 1`] = `
{
  "bBottom": 10,
  "bLeft": 20,
  "bRight": 20,
  "bTop": 10,
}
`;

exports[`Get Css Sides computes full sides 1`] = `
{
  "dBottom": 30,
  "dLeft": 40,
  "dRight": 20,
  "dTop": 10,
}
`;

exports[`Get Css Sides computes number padding 1`] = `
{
  "f": 40,
}
`;

exports[`Get Css Sides computes single side 1`] = `
{
  "aBottom": 10,
  "aLeft": 10,
  "aRight": 10,
  "aTop": 10,
}
`;

exports[`Get Css Sides computes tripple sides 1`] = `
{
  "cBottom": 30,
  "cLeft": 20,
  "cRight": 20,
  "cTop": 10,
}
`;

exports[`Get Css Sides returns null if empty string 1`] = `null`;

exports[`Get Css Sides returns null if no padding 1`] = `null`;

exports[`Get Css Sides returns null if the name is empty 1`] = `null`;

exports[`Get Css Sides trims extra sides 1`] = `
{
  "eBottom": 30,
  "eLeft": 40,
  "eRight": 20,
  "eTop": 10,
}
`;
